service/test/_amqp_test.js

```diff
@@ -134,69 +156,542 @@ describe('Amqp', function() {
-        assert(putTokenStub.calledOnce);
-        assert(putTokenStub.calledWith('uri',sasConfig.sharedAccessSignature.toString()));
-        assert(disconnectStub.calledOnce);
+        assert(fakeAmqpBase.putToken.calledOnce);
+        assert(fakeAmqpBase.putToken.calledWith('uri',sasConfig.sharedAccessSignature.toString()));
+        assert(fakeAmqpBase.disconnect.calledOnce);
         assert.instanceOf(err, Error);
       });
     });
 
+    it('calls its callback immediately if it is already connected', function (testCallback) {
+      var transport = new Amqp(sasConfig, fakeAmqpBase);
+      transport.connect(function () {
+        assert.isTrue(fakeAmqpBase.connect.calledOnce);
+        assert.isTrue(fakeAmqpBase.initializeCBS.calledOnce);
+        assert.isTrue(fakeAmqpBase.putToken.calledOnce);
+        transport.connect(function () {
+          assert.isTrue(fakeAmqpBase.connect.calledOnce);
+          assert.isTrue(fakeAmqpBase.initializeCBS.calledOnce);
+          assert.isTrue(fakeAmqpBase.putToken.calledOnce);
+          testCallback();
+        });
+      });
+    });
   });
 
   describe('#disconnect', function() {
     /*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_020: [** The `disconnect` method shall call the `disconnect` method of the base AMQP transport and translate its result to the caller into a transport-agnostic object.]*/
-    it('calls the base transport disconnect method', function(done) {
-      var amqp = new Amqp(fakeConfig, new FakeAmqpBase());
-      amqp.disconnect(done);
+    it('calls the base transport disconnect method if the transport is connected', function(testCallback) {
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
+      amqp.connect(function () {
+        amqp.disconnect(function () {
+          assert.isTrue(fakeAmqpBase.disconnect.calledOnce);
+          testCallback();
+        });
+      });
+    });
+
+    /*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_020: [** The `disconnect` method shall call the `disconnect` method of the base AMQP transport and translate its result to the caller into a transport-agnostic object.]*/
+    it('calls the base transport disconnect method if the transport is connecting', function(testCallback) {
+      fakeAmqpBase.connect = sinon.stub();
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
+      amqp.connect(function () {});
+      // blocked in connecting because the connect callback won't be called.
+      amqp.disconnect(function () {
+        assert.isTrue(fakeAmqpBase.disconnect.calledOnce);
+        testCallback();
+      });
+    });
+
+    /*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_018: [All asynchronous instance methods shall call the `done` callback with either no arguments or a first null argument and a second argument that is the result of the operation if the operation succeeded.]*/
+    it('calls the base transport disconnect method if the transport is authenticating', function(testCallback) {
+      fakeAmqpBase.putToken = sinon.stub();
```

Comment:
>fakeAmqpBase.putToken = sinon.stub(); [](start = 6, length = 37)

It's unclear to me what this has to do with the test.

Reply:
I'm replacing the existing stub of putToken (existing stub calls its callback) with a stub() that does nothing. therefore it blocks the state machine in the "authenticating" state.

---
In reply to: [147616289](https://github.com/Azure/azure-iot-sdk-node/pull/127#discussion_r147616289) [](ancestors = 147616289)